kernel/src/device/fb.rs

```diff
@@ -72,20 +73,52 @@ impl FileIo for Fb {
 }
 
 impl FileIo for FbHandle {
-    fn read(&self, _writer: &mut VmWriter) -> Result<usize> {
-        // TODO: Add support for reading from the framebuffer using `read`
-        return_errno_with_message!(
-            Errno::EOPNOTSUPP,
-            "reading from the framebuffer is not supported"
-        );
+    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
+        let buffer_size = self.framebuffer.io_mem().size();
+        let mut offset = self.offset.lock();
+
+        if *offset >= buffer_size {
+            return Ok(0); // EOF
+        }
+
+        let read_len = writer.avail().min(buffer_size - *offset);
+        if read_len == 0 {
+            return Ok(0);
+        }
+
+        // Read from the framebuffer at the current offset.
+        // Limit the writer to avoid over-reading when the user buffer is
+        // larger than the remaining framebuffer size.
+        self.framebuffer
+            .io_mem()
+            .read(*offset, writer.limit(read_len))?;
+
+        *offset += read_len;
+        Ok(read_len)
     }
 
-    fn write(&self, _reader: &mut VmReader) -> Result<usize> {
-        // TODO: Add support for writing to the framebuffer using `write`
-        return_errno_with_message!(
-            Errno::EOPNOTSUPP,
-            "writing to the framebuffer is not supported"
-        );
+    fn write(&self, reader: &mut VmReader) -> Result<usize> {
+        let buffer_size = self.framebuffer.io_mem().size();
+        let mut offset = self.offset.lock();
+
+        if *offset >= buffer_size {
+            return_errno_with_message!(Errno::ENOSPC, "write beyond framebuffer size");
```

Comment:
```suggestion
            return_errno_with_message!(Errno::ENOSPC, "the write offset is beyoud the framebuffer size");
```

Are you sure this shouldn't be an `OK (0)`?

Reply:
I believe that when writing to the framebuffer beyond its capacity, the return value should be `ENOSPC`. I have tested this on Linux and included it in the CI test :-)